fecgov/openFEC
# Fix candidate missing from candidate totals

## Summary (required)

- Resolves #3768 and partial resolution for #3709 
ofec_candidate_totals_mv has two major components, one from financial side, one from the candidate site.
There are two major types of candidates that caused the mismatch of these two worlds. 
- candidates has committees but did not file financial reports
- candidates has no committees at all

We had noticed the candidate_election_year logic has problem.
In previous ticket #3700, we had fixed the financial side.  
In another ticket #3736, the candidate side had been fixed.
This ticket we combined the two sides together with some minor updates
we noticed from previous tickets.

## How to test the changes locally

-Download the branch feature/3768-candidate-missing-from-candidate-totals, run migration file V0136 through V0139 (or run invoke create_sample_db) to make sure migration files runs without error.

tmp materialized views had been created in DEV database that has the updated MVs logic.
(Note: public.ofec_candidate_history_with_future_election_mv_tmp_jj2 does not change in this ticket, but it references public.ofec_candidate_history_mv_tmp_jj2)
public.ofec_candidate_history_mv_tmp_jj2
public.ofec_candidate_history_with_future_election_mv_tmp_jj2
public.ofec_cand_cmte_linkage_mv_tmp_jj2
public.ofec_candidate_totals_mv_tmp_jj2
public.ofec_candidate_detail_mv_tmp_jj2

To test database:
(Cand_id:
-- 'P00003392': Hillary Clinton (runs 2008, 2016)
make sure 2014 cycle financial data is calculated as part of election_cycle 2016, not 2008; 2010, 2012, 2018, and 2020 cycle data should not be calculated as part of neither election_cycle 2008 nor 2016.
-- 'H6PR00082':  (runs 2016, 2020)
Puerto Rico's resident commissioner has a 4-year election cycle. Even though the commissioner serves in the house.  Make sure the 2018 cycle data also been included in 2020 election_cycle, even this candidate serves in house.
 
--Connect to DEV database, testing any of the following examples.
select cand_id, cand_election_yr, fec_election_yr, election_yr_to_be_included 
from ofec_cand_cmte_linkage_mv_tmp_jj2 
where cand_id = 'H6PR00082'
order by cand_election_yr, fec_election_yr, election_yr_to_be_included:

select * from ofec_candidate_totals_mv_tmp_jj2 
where candidate_id = 'H6PR00082'
order by election_year, cycle, is_election;

-- Check candidate list and candidate totals has difference in number of candidates for election_year 2020
-- Note:  candidate _detail_mv has real election_year while candidate_totals_mv has rounded election_year to match the fec_election_year, so to compare 2020 cycle, array 2019, 2020 need to be used in candidate_detail_mv.

(
select 'totals', candidate_id from ofec_candidate_totals_mv_tmp_jj2 where election_year= 2020 and is_election = true and candidate_inactive = false
except 
select 'totals', detail.candidate_id
FROM ofec_candidate_detail_mv_tmp_jj2 detail
WHERE detail.election_years && Array[2019,2020]
and candidate_inactive = False
)
union
(
select 'cand', detail.candidate_id
FROM ofec_candidate_detail_mv_tmp_jj2 detail
WHERE detail.election_years && Array[2019,2020]
and candidate_inactive = False
except
select 'cand', candidate_id from ofec_candidate_totals_mv_tmp_jj2 where election_year= 2020 and is_election = true and candidate_inactive = false
)
order by 2, 1;

- To test API:
On your local machine
In **models/candidates.py** file, make the following changes:
In class CandidateHistory(BaseCandidate):
tablename = 'ofec_candidate_history_mv_tmp_jj2'

In class CandidateHistoryWithFuture(BaseCandidate):
tablename = 'ofec_candidate_history_with_future_election_mv_tmp_jj2'

In class CandidateTotal(db.Model):
tablename = 'ofec_candidate_totals_mv_tmp_jj2'

In class Candidate:
replace ofec_cand_cmte_linkage_mv with ofec_cand_cmte_linkage_mv_tmp_jj2 in
secondaryjoin=
secondaryjoin=
order_by=

In class BaseConcreteCandidate:
tablename = 'ofec_candidate_detail_mv_tmp_jj2'

In class Candidate:
tablename = 'ofec_candidate_detail_mv_tmp_jj2'

In class CandidateDetail:
tablename = 'ofec_candidate_detail_mv_tmp_jj2'

In **models/links.py** file, make the following changes:
In class CandidateCommitteeLink:
tablename = 'ofec_cand_cmte_linkage_mv_tmp_jj2'

candidate_id = db.Column('cand_id', db.String, db.ForeignKey('ofec_candidate_detail_mv_tmp_jj2.candidate_id'))

Point to the DEV database and start the local server.
For /candidate/totals endpoint:
election_year: 2020
office: P
election_full: true
(noted the count)

candidate_id: 'H6PR00082'
election_year: 2020
election_full: true
(noted the receipts and disbursements) 

For /candidates/ endpoint:
election_year: 2019, 2020
office: P
(noted the count.  The count should match the count from /candidate/totals endpoint)

## Impacted areas of the application
List general components of the application that this PR will affect:

There is no API change in this ticket. Output data may change due to the underline MV changes.

/candidate/{candidate_id}/history/
/candidates/totals/
/candidates/

## Related PRs
List related PRs against other branches:

branch | PR
------ | ------
fix/other_pr | [link]()
feature/other_pr | [link]()
